pkg/agent/client/client.go

```diff
@@ -0,0 +1,213 @@
+		tlsConfig = spiffePeer.NewTLSConfig(tlsCert)
+		return credentials.NewTLS(tlsConfig), nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second) // TODO: Make this timeout configurable?
+	defer cancel()
+
+	config := grpcutil.GRPCDialerConfig{
+		Log:      grpcutil.LoggerFromFieldLogger(c.Log),
+		CredFunc: credFunc,
+	}
+	dialer := grpcutil.NewGRPCDialer(config)
+	conn, err := dialer.Dial(ctx, c.Addr)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create connection: %v", err)
+	}
+	return conn, nil
+}
+
+func (c *client) FetchUpdates(req *node.FetchSVIDRequest) (*Update, error) {
+	nodeClient, err := c.newNodeClient()
+	if err != nil {
+		return nil, err
+	}
+
+	stream, err := nodeClient.FetchSVID(context.Background())
+	// We weren't able to get a stream...close the client and return the error.
+	if err != nil {
+		c.Release()
+		c.c.Log.Errorf("%v: %v", ErrUnableToGetStream, err)
+		return nil, ErrUnableToGetStream
+	}
+
+	// Send the request to the server using the stream.
+	err = stream.Send(req)
+	// Close the stream whether there was an error or not
+	stream.CloseSend()
+	if err != nil {
+		// TODO: should we try to create a new stream?
+		return nil, err
+	}
+
+	regEntries := map[string]*common.RegistrationEntry{}
+	svids := map[string]*node.Svid{}
+	var lastBundle []byte
+	// Read all the server responses from the stream.
+	for {
+		resp, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			if len(regEntries) > 0 || len(svids) > 0 || lastBundle != nil {
+				// There was an error receiving a response, exit loop to return what we have.
+				return &Update{regEntries, svids, lastBundle}, ErrPartialResponse
+			}
+
+			return nil, err
+		}
+
+		for _, re := range resp.SvidUpdate.RegistrationEntries {
+			regEntries[re.EntryId] = re
+		}
+		for spiffeid, svid := range resp.SvidUpdate.Svids {
+			svids[spiffeid] = svid
+		}
+		lastBundle = resp.SvidUpdate.Bundle
+	}
+	return &Update{regEntries, svids, lastBundle}, nil
+}
+
+func (c *client) Release() {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.conn != nil {
+		c.conn.Close()
+		c.conn = nil
+	}
+}
+
+func (c *client) newNodeClient() (node.NodeClient, error) {
+	if c.newNodeClientCallback != nil {
+		return c.newNodeClientCallback()
+	}
+
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.conn == nil {
+		conn, err := newGRPCConn(c.c)
+		if err != nil {
+			return nil, err
+		}
+		c.conn = conn
+	}
+	return node.NewNodeClient(c.conn), nil
+}
+
+func (u *Update) String() string {
+	var buffer bytes.Buffer
+	buffer.WriteString("{ Entries: [")
+	for _, re := range u.Entries {
+		buffer.WriteString("{ spiffeID: ")
+		buffer.WriteString(re.SpiffeId)
+		buffer.WriteString(", parentID: ")
+		buffer.WriteString(re.ParentId)
+		buffer.WriteString(", selectors: ")
+		buffer.WriteString(fmt.Sprintf("%v", re.Selectors))
+		buffer.WriteString("}")
+	}
+	buffer.WriteString("], SVIDs: [")
+	for key, svid := range u.SVIDs {
+		buffer.WriteString(key)
```

Comment:
Feels dangerous to have code hanging around that will print private keys. Maybe it's better to omit the keys? Or print `[REDACTED]` or something similar?

Reply:
I think I didn't choose a happy name for that variable...it is just the key of the SVIDs map, which is a spiffe id. Maybe I could rename it to spiffeID?

Reply:
i think the "key" here is the spiffeid....